core/services/keeper/registry_interface.go

```diff
@@ -0,0 +1,245 @@
+package keeper
+
+import (
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/pkg/errors"
+	evmclient "github.com/smartcontractkit/chainlink/core/chains/evm/client"
+	registry1_1 "github.com/smartcontractkit/chainlink/core/internal/gethwrappers/generated/keeper_registry_wrapper1_1"
+	registry1_2 "github.com/smartcontractkit/chainlink/core/internal/gethwrappers/generated/keeper_registry_wrapper1_2"
+	"github.com/smartcontractkit/chainlink/core/services/keystore/keys/ethkey"
+)
+
+type RegistryVersion int32
+
+const (
+	RegistryVersion_1_0 RegistryVersion = iota
+	RegistryVersion_1_1
+	RegistryVersion_1_2
+)
+
+// RegistryWrapper implements a layer on top of different versions of registry wrappers
+// to provide a unified layer to rest of the codebase
+type RegistryWrapper struct {
+	Address     ethkey.EIP55Address
+	Version     RegistryVersion
+	contract1_1 *registry1_1.KeeperRegistry
+	contract1_2 *registry1_2.KeeperRegistry
+}
+
+func NewRegistryWrapper(address ethkey.EIP55Address, backend bind.ContractBackend) (*RegistryWrapper, error) {
+	contract1_1, err := registry1_1.NewKeeperRegistry(
+		address.Address(),
+		backend,
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to create keeper registry 1_1 contract wrapper")
+	}
+	contract1_2, err := registry1_2.NewKeeperRegistry(
+		address.Address(),
+		backend,
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to create keeper registry 1_2 contract wrapper")
+	}
+
+	version, err := getRegistryVersion(contract1_1)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to determine version of keeper registry contract")
+	}
+
+	return &RegistryWrapper{
+		Address:     address,
+		Version:     *version,
+		contract1_1: contract1_1,
+		contract1_2: contract1_2,
+	}, nil
+}
+
+func getRegistryVersion(contract1_1 *registry1_1.KeeperRegistry) (*RegistryVersion, error) {
```

Comment:
Shouldn't this be like...
```go
func getRegistryVersion(contract *type_and_version.TypeAndVersion) (*RegistryVersion, error) {
```

Reply:
ah interesting suggestion, I'll try generating gethwrapper for typeAndVersion and see if this works

Reply:
done